packages/core/shared/process-chat-stream.ts

```diff
@@ -56,32 +86,129 @@ export async function processChatStream({
+          );
+          continue;
+        }
         const functionCallResponse: ChatRequest | void =
           await experimental_onFunctionCall(getCurrentMessages(), functionCall);
 
         // If the user does not return anything as a result of the function call, the loop will break.
         if (functionCallResponse === undefined) break;
         // A function call response was returned.
         // The updated chat with function call response will be sent to the API in the next iteration of the loop.
+        fixFunctionCallArguments(functionCallResponse);
         updateChatRequest(functionCallResponse);
       }
+      // If we get here and are expecting a tool call, the message should have one, if not warn and continue
+      if (experimental_onToolCall) {
+        const toolCalls = streamedResponseMessage.tool_calls;
+        if (!(typeof toolCalls === 'object')) {
+          console.warn(
+            'experimental_onToolCall should not be defined when using functions',
+          );
+          continue;
+        }
+        const toolCallResponse: ChatRequest | void =
+          await experimental_onToolCall(getCurrentMessages(), toolCalls);
+
+        // If the user does not return anything as a result of the function call, the loop will break.
+        if (toolCallResponse === undefined) break;
+        // A function call response was returned.
+        // The updated chat with function call response will be sent to the API in the next iteration of the loop.
+        fixFunctionCallArguments(toolCallResponse);
+        updateChatRequest(toolCallResponse);
+      }
+
+      // Make sure funtion call arguments are sent back to the API as a string
+      function fixFunctionCallArguments(response: ChatRequest) {
+        for (const message of response.messages) {
+          if (message.tool_calls !== undefined) {
+            for (const toolCall of message.tool_calls) {
+              if (typeof toolCall === 'object') {
+                if (
+                  toolCall.function.arguments &&
+                  typeof toolCall.function.arguments !== 'string'
+                ) {
+                  toolCall.function.arguments = JSON.stringify(
+                    toolCall.function.arguments,
+                  );
+                }
+              }
+            }
+          }
+          if (message.function_call !== undefined) {
+            if (typeof message.function_call === 'object') {
+              if (
+                message.function_call.arguments &&
+                typeof message.function_call.arguments !== 'string'
+              ) {
+                message.function_call.arguments = JSON.stringify(
+                  message.function_call.arguments,
+                );
+              }
+            }
+          }
+        }
+      }
     }
   }
+  if (count >= maxIterations) {
+    console.error(
+      'Max iterations reached for processChatStream probably an infinite ly an infinite loop.',
+    );
+  }
```

Comment:
This got me thinking, why would this happen for tools but not for functions? It could be many things:

1. missing test for function
2. something different in the tools test
3. something different between functions / tools

Can the infinite loop (that's not infinite any more) happen outside test situations? why does it happen?